db/mongodb/mongodb.go

```diff
@@ -0,0 +1,221 @@
+	clientOptions := options.Client().ApplyURI(url).SetMaxConnecting(20)
+	ctx, cancel := context.WithTimeout(context.Background(), MONGODB_TIMEOUT_CONNECT)
+	defer cancel()
+	client, err := mongo.Connect(ctx, clientOptions)
+	if err != nil {
+		return nil, err
+	}
+	database := client.Database(collection)
+
+	// Shutdown database connection when SIGTERM received
+	go func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+		<-c
+		log.Warnf("received SIGTERM, disconnecting mongo database")
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		err := client.Disconnect(ctx)
+		if err != nil {
+			log.Warn(err)
+		}
+		cancel()
+	}()
+
+	err = client.Ping(ctx, readpref.Primary())
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to mongodb: %w", err)
+	}
+
+	return &MongoDB{db: database, collection: collection}, nil
+}
+
+func (d *MongoDB) Close() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	return d.db.Client().Disconnect(ctx)
+}
+
+func (d *MongoDB) WriteTx() db.WriteTx {
+	return &WriteTx{
+		db:         d.db,
+		batch:      make([]mongo.WriteModel, 0),
+		inMem:      make(map[string][]byte),
+		collection: d.collection,
+	}
+}
+
+func (d *MongoDB) Get(key []byte) ([]byte, error) {
+	tx := d.WriteTx()
+	defer tx.Discard()
+	return tx.Get(key)
+}
+
+func (d *MongoDB) Iterate(prefix []byte, callback func(key, value []byte) bool) error {
+	tx := d.WriteTx()
+	defer tx.Discard()
+	return tx.Iterate(prefix, callback)
+}
+
+func (d *MongoDB) Compact() error {
+	// MongoDB handles compaction automatically, so this is a no-op.
+	return nil
+}
+
+// check that WriteTx implements the db.WriteTx interface
+var _ db.WriteTx = (*WriteTx)(nil)
+
+type WriteTx struct {
+	db         *mongo.Database
+	batch      []mongo.WriteModel
+	inMem      map[string][]byte
+	collection string
+}
+
+func (tx *WriteTx) Get(k []byte) ([]byte, error) {
+	if val, ok := tx.inMem[string(k)]; ok {
+		return val, nil
+	}
+	collection := tx.db.Collection(tx.collection)
+
+	filter := bson.M{"_id": string(k)} // Convert to string
+	var result KeyVal
+	ctx, cancel := context.WithTimeout(context.Background(), MONGODB_TIMEOUT_QUERY)
+	defer cancel()
+	err := collection.FindOne(ctx, filter).Decode(&result)
+	if err == mongo.ErrNoDocuments {
+		return nil, db.ErrKeyNotFound
+	}
+	return result.Value, err
+}
+
+func (tx *WriteTx) Iterate(prefix []byte, callback func(key, value []byte) bool) error {
+	// First, handle in-memory keys
+	for k, v := range tx.inMem {
+		if strings.HasPrefix(k, string(prefix)) {
+			if !callback([]byte(k), v) {
+				return nil
+			}
+		}
+	}
+
+	// Now, handle database keys, but skip those already handled from in-memory map
+	collection := tx.db.Collection(tx.collection)
+
+	filter := bson.M{}
+	if len(prefix) > 0 {
+		filter = bson.M{"_id": bson.M{"$regex": primitive.Regex{Pattern: "^" + string(prefix)}}}
```

Comment:
I wonder how slow this might be...

Reply:
I hope mongo people made it efficient enough :/ But there is not other way, I have been looking for a better (non string regex bassed) mechanism, but there is nothing we can use.